Extended client not work when at least one api is not available in kubernetes cluster.

**Describe the bug**
When using extended client to interact with a kubernetes cluster, it will first try to refresh model mapping, for this it will list all available apis in cluster, and then call each of them. When at least one of these apis is unavailable, an exception is thrown and client call is not executed : 
```
io.kubernetes.client.extended.kubectl.exception.KubectlException: io.kubernetes.client.openapi.ApiException:
        at io.kubernetes.client.extended.kubectl.Kubectl$ApiClientBuilder.refreshDiscovery(Kubectl.java:238)
        at io.kubernetes.client.extended.kubectl.KubectlGet.execute(KubectlGet.java:60)
        ...
Caused by: io.kubernetes.client.openapi.ApiException:
        at io.kubernetes.client.openapi.ApiClient.handleResponse(ApiClient.java:973)
        at io.kubernetes.client.openapi.ApiClient.execute(ApiClient.java:885)
        at io.kubernetes.client.Discovery.get(Discovery.java:174)
        at io.kubernetes.client.Discovery.resourceDiscovery(Discovery.java:156)
        at io.kubernetes.client.Discovery.findAll(Discovery.java:71)
        at io.kubernetes.client.Discovery.findAll(Discovery.java:65)
        at io.kubernetes.client.Discovery.findAll(Discovery.java:55)
        at io.kubernetes.client.util.ModelMapper.refresh(ModelMapper.java:260)
        at io.kubernetes.client.util.ModelMapper.refresh(ModelMapper.java:240)
        at io.kubernetes.client.extended.kubectl.Kubectl$ApiClientBuilder.refreshDiscovery(Kubectl.java:236)
```

**Client Version**
`18.0.0`

**To Reproduce**
Register an api in error in kubernetes cluster: 
```
cat << EOF | kubectl apply -f -
apiVersion: apiregistration.k8s.io/v1
kind: APIService
metadata:
  name: v1alpha1.bug.reproduce.java-client
spec:
  service:
    name: not-existing-service
    namespace: default
  group: bug.reproduce.java-client
  version: v1alpha1
  insecureSkipTLSVerify: true
  groupPriorityMinimum: 10
  versionPriority: 10
EOF
```
Use the extended client: `Kubectl.get(V1Deployment.class).execute()`

It will fail because the previously created apiService is not available.

**Expected behavior**
The extended client call should be executed even if another api is unavailable in cluster

--- a/extended/src/main/java/io/kubernetes/client/extended/kubectl/KubectlApiResources.java
+++ b/extended/src/main/java/io/kubernetes/client/extended/kubectl/KubectlApiResources.java
@@ -15,16 +15,24 @@ package io.kubernetes.client.extended.kubectl;
 import io.kubernetes.client.Discovery;
 import io.kubernetes.client.extended.kubectl.exception.KubectlException;
 import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.util.exception.IncompleteDiscoveryException;
 import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class KubectlApiResources extends Kubectl.ApiClientBuilder<KubectlApiResources>
     implements Kubectl.Executable<Set<Discovery.APIResource>> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(KubectlApiResources.class);
 
   @Override
   public Set<Discovery.APIResource> execute() throws KubectlException {
     Discovery discovery = new Discovery(this.apiClient);
     try {
       return discovery.findAll();
+    } catch (IncompleteDiscoveryException e) {
+      LOGGER.warn("Error while getting all api resources, some resources will be missing", e);
+      return e.getDiscoveredResources();
     } catch (ApiException e) {
       throw new KubectlException(e);
     }

--- a/util/src/main/java/io/kubernetes/client/Discovery.java
+++ b/util/src/main/java/io/kubernetes/client/Discovery.java
@@ -72,7 +72,7 @@ public class Discovery {
       V1APIResourceList resourceList = resourceDiscovery(path);
       return groupResourcesByName(group, versions, preferredVersion, resourceList);
     } catch (ApiException e) {
-      LOGGER.warn("Api {} returns a {} code, all api resources managed by this api cannot be discovered", path, e.getCode(), e);
+      LOGGER.warn("Unable to retrieve the complete list of server APIs: {}/{}: {}", group, preferredVersion, e.getResponseBody());
       return Collections.emptySet();
     }
   }

--- a/util/src/main/java/io/kubernetes/client/Discovery.java
+++ b/util/src/main/java/io/kubernetes/client/Discovery.java
@@ -20,6 +20,7 @@ import io.kubernetes.client.openapi.models.V1APIGroup;
 import io.kubernetes.client.openapi.models.V1APIGroupList;
 import io.kubernetes.client.openapi.models.V1APIResourceList;
 import io.kubernetes.client.openapi.models.V1APIVersions;
+import io.kubernetes.client.util.exception.IncompleteDiscoveryException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
@@ -32,12 +33,9 @@ import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
 import okhttp3.Call;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 public class Discovery {
 
-  private static final Logger LOGGER = LoggerFactory.getLogger(Discovery.class);
   private final ApiClient apiClient;
 
   public Discovery() {
@@ -53,28 +51,42 @@ public class Discovery {
     for (String version : legacyCoreApi().getVersions()) {
       allResources.addAll(findAll("", Arrays.asList(version), version, "/api/" + version));
     }
+    IncompleteDiscoveryException incompleteDiscoveryException = null;
     for (V1APIGroup group : groupDiscovery("/apis").getGroups()) {
-      allResources.addAll(
-          findAll(
-              group.getName(),
-              group.getVersions().stream().map(v -> v.getVersion()).collect(Collectors.toList()),
-              group.getPreferredVersion().getVersion()));
+      try {
+        allResources.addAll(
+            findAll(
+                group.getName(),
+                group.getVersions().stream().map(v -> v.getVersion()).collect(Collectors.toList()),
+                group.getPreferredVersion().getVersion()));
+      } catch (ApiException e){
+        IncompleteDiscoveryException resourceDiscoveryException = new IncompleteDiscoveryException(
+            String.format("Unable to retrieve the complete list of server APIs: %s/%s : %s",
+                group.getName(), group.getPreferredVersion().getVersion(), e.getResponseBody()),
+            e, allResources);
+        if (incompleteDiscoveryException == null) {
+          incompleteDiscoveryException = resourceDiscoveryException;
+        }else {
+          incompleteDiscoveryException.addSuppressed(resourceDiscoveryException);
+        }
+      }
+    }
+    if (incompleteDiscoveryException != null) {
+      throw incompleteDiscoveryException;
     }
     return allResources;
   }
 
-  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion) {
+  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion)
+      throws ApiException {
     return findAll(group, versions, preferredVersion, "/apis/" + group + "/" + preferredVersion);
   }
 
-  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion, String path) {
-    try {
-      V1APIResourceList resourceList = resourceDiscovery(path);
-      return groupResourcesByName(group, versions, preferredVersion, resourceList);
-    } catch (ApiException e) {
-      LOGGER.warn("Unable to retrieve the complete list of server APIs: {}/{}: {}", group, preferredVersion, e.getResponseBody());
-      return Collections.emptySet();
-    }
+  public Set<APIResource> findAll(
+      String group, List<String> versions, String preferredVersion, String path)
+      throws ApiException {
+    V1APIResourceList resourceList = resourceDiscovery(path);
+    return groupResourcesByName(group, versions, preferredVersion, resourceList);
   }
 
   public Set<APIResource> groupResourcesByName(

--- a/util/src/main/java/io/kubernetes/client/Discovery.java
+++ b/util/src/main/java/io/kubernetes/client/Discovery.java
@@ -32,9 +32,12 @@ import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
 import okhttp3.Call;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class Discovery {
 
+  private static final Logger LOGGER = LoggerFactory.getLogger(Discovery.class);
   private final ApiClient apiClient;
 
   public Discovery() {
@@ -60,16 +63,18 @@ public class Discovery {
     return allResources;
   }
 
-  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion)
-      throws ApiException {
+  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion) {
     return findAll(group, versions, preferredVersion, "/apis/" + group + "/" + preferredVersion);
   }
 
-  public Set<APIResource> findAll(
-      String group, List<String> versions, String preferredVersion, String path)
-      throws ApiException {
-    V1APIResourceList resourceList = resourceDiscovery(path);
-    return groupResourcesByName(group, versions, preferredVersion, resourceList);
+  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion, String path) {
+    try {
+      V1APIResourceList resourceList = resourceDiscovery(path);
+      return groupResourcesByName(group, versions, preferredVersion, resourceList);
+    } catch (ApiException e) {
+      LOGGER.warn("Api {} returns a {} code, all api resources managed by this api cannot be discovered", path, e.getCode(), e);
+      return Collections.emptySet();
+    }
   }
 
   public Set<APIResource> groupResourcesByName(

--- a/util/src/main/java/io/kubernetes/client/util/ModelMapper.java
+++ b/util/src/main/java/io/kubernetes/client/util/ModelMapper.java
@@ -18,6 +18,7 @@ import io.kubernetes.client.apimachinery.GroupVersionResource;
 import io.kubernetes.client.common.KubernetesListObject;
 import io.kubernetes.client.common.KubernetesObject;
 import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.util.exception.IncompleteDiscoveryException;
 import java.io.File;
 
 import java.io.IOException;
@@ -266,7 +267,13 @@ public class ModelMapper {
       return lastAPIDiscovery;
     }
 
-    Set<Discovery.APIResource> apiResources = discovery.findAll();
+    Set<Discovery.APIResource> apiResources = null;
+    try {
+      apiResources = discovery.findAll();
+    } catch (IncompleteDiscoveryException e) {
+      logger.warn("Error while getting all api resources, some api resources will not be refreshed", e);
+      apiResources = e.getDiscoveredResources();
+    }
 
     for (Discovery.APIResource apiResource : apiResources) {
       for (String version : apiResource.getVersions()) {

--- a/util/src/main/java/io/kubernetes/client/util/exception/IncompleteDiscoveryException.java
+++ b/util/src/main/java/io/kubernetes/client/util/exception/IncompleteDiscoveryException.java
@@ -1,3 +1,15 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
 package io.kubernetes.client.util.exception;
 
 import io.kubernetes.client.Discovery.APIResource;

--- a/util/src/main/java/io/kubernetes/client/util/exception/IncompleteDiscoveryException.java
+++ b/util/src/main/java/io/kubernetes/client/util/exception/IncompleteDiscoveryException.java
@@ -1,0 +1,22 @@
+package io.kubernetes.client.util.exception;
+
+import io.kubernetes.client.Discovery.APIResource;
+import io.kubernetes.client.openapi.ApiException;
+import java.util.Set;
+
+public class IncompleteDiscoveryException extends ApiException {
+
+  private static final long serialVersionUID = 1L;
+
+  private final transient Set<APIResource> discoveredResources;
+
+  public IncompleteDiscoveryException(String message, ApiException cause, Set<APIResource> discoveredResources) {
+    super(message, cause, cause.getCode(), cause.getResponseHeaders(), cause.getResponseBody());
+    this.discoveredResources = discoveredResources;
+  }
+
+  public Set<APIResource> getDiscoveredResources() {
+    return discoveredResources;
+  }
+
+}

--- a/util/src/test/java/io/kubernetes/client/DiscoveryTest.java
+++ b/util/src/test/java/io/kubernetes/client/DiscoveryTest.java
@@ -20,6 +20,7 @@ import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.options
 import static org.junit.Assert.assertEquals;
 
 import com.github.tomakehurst.wiremock.junit.WireMockRule;
+import io.kubernetes.client.Discovery.APIResource;
 import io.kubernetes.client.openapi.ApiClient;
 import io.kubernetes.client.openapi.ApiException;
 import io.kubernetes.client.openapi.models.V1APIResource;
@@ -28,6 +29,7 @@ import io.kubernetes.client.openapi.models.V1APIVersions;
 import io.kubernetes.client.util.ClientBuilder;
 import java.io.IOException;
 import java.util.Arrays;
+import java.util.List;
 import java.util.Set;
 import org.junit.Before;
 import org.junit.Rule;
@@ -100,5 +102,54 @@ public class DiscoveryTest {
     assertEquals("meow", meow.getResourceSingular());
     assertEquals(true, meow.getNamespaced());
     assertEquals("mouse", meow.getSubResources().get(0));
+  }
+
+  @Test
+  public void findAllShouldReturnAllApiResourceWhenApiResponseIsSuccess() {
+    Discovery discovery = new Discovery(apiClient);
+
+    String group = "test.com";
+    String version = "v1";
+    String path="/apis/"+group+'/'+version;
+
+    wireMockRule.stubFor(
+        get(urlPathEqualTo(path))
+            .willReturn(
+                aResponse()
+                    .withStatus(200)
+                    .withHeader("Content-Type", "application/json")
+                    .withBody(
+                        apiClient
+                            .getJSON()
+                            .serialize(new V1APIResourceList()
+                                .resources(
+                                    Arrays.asList(
+                                      new V1APIResource()
+                                        .name("first"),
+                                      new V1APIResource()
+                                        .name("second")))))));
+
+    Set<APIResource> apiResources = discovery.findAll(group, List.of(version), version);
+    wireMockRule.verify(1, getRequestedFor(urlPathEqualTo(path)));
+    assertEquals(2, apiResources.size());
+  }
+
+  @Test
+  public void findAllShouldReturnEmptyListWhenApiResponseIsNotSuccess() {
+    Discovery discovery = new Discovery(apiClient);
+
+    String group = "test.com";
+    String version = "v1";
+    String path="/apis/"+group+'/'+version;
+
+    wireMockRule.stubFor(
+        get(urlPathEqualTo(path))
+            .willReturn(
+                aResponse()
+                    .withStatus(503)));
+
+    Set<APIResource> apiResources = discovery.findAll(group, List.of(version), version);
+    wireMockRule.verify(1, getRequestedFor(urlPathEqualTo(path)));
+    assertEquals(0, apiResources.size());
   }
 }

--- a/util/src/test/java/io/kubernetes/client/DiscoveryTest.java
+++ b/util/src/test/java/io/kubernetes/client/DiscoveryTest.java
@@ -18,15 +18,20 @@ import static com.github.tomakehurst.wiremock.client.WireMock.getRequestedFor;
 import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
 import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.options;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
 
 import com.github.tomakehurst.wiremock.junit.WireMockRule;
 import io.kubernetes.client.Discovery.APIResource;
 import io.kubernetes.client.openapi.ApiClient;
 import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.openapi.models.V1APIGroup;
+import io.kubernetes.client.openapi.models.V1APIGroupList;
 import io.kubernetes.client.openapi.models.V1APIResource;
 import io.kubernetes.client.openapi.models.V1APIResourceList;
 import io.kubernetes.client.openapi.models.V1APIVersions;
+import io.kubernetes.client.openapi.models.V1GroupVersionForDiscovery;
 import io.kubernetes.client.util.ClientBuilder;
+import io.kubernetes.client.util.exception.IncompleteDiscoveryException;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -106,12 +111,39 @@ public class DiscoveryTest {
   }
 
   @Test
-  public void findAllShouldReturnAllApiResourceWhenApiResponseIsSuccess() {
+  public void findAllShouldReturnAllApiResourceWhenAllResourceDiscoveryApiResponseAreSuccess() throws ApiException {
     Discovery discovery = new Discovery(apiClient);
+
+    wireMockRule.stubFor(
+        get("/api")
+            .willReturn(
+                aResponse()
+                    .withStatus(200)
+                    .withHeader("Content-Type", "application/json")
+                    .withBody(
+                        apiClient
+                            .getJSON()
+                            .serialize(new V1APIVersions()))));
 
     String group = "test.com";
     String version = "v1";
     String path="/apis/"+group+'/'+version;
+
+    wireMockRule.stubFor(
+        get("/apis")
+            .willReturn(
+                aResponse()
+                    .withStatus(200)
+                    .withHeader("Content-Type", "application/json")
+                    .withBody(
+                        apiClient
+                            .getJSON()
+                            .serialize(new V1APIGroupList()
+                                .addGroupsItem(new V1APIGroup()
+                                    .name(group)
+                                    .preferredVersion(new V1GroupVersionForDiscovery().version(version))
+                                    .versions(Arrays.asList(
+                                        new V1GroupVersionForDiscovery().version(version))))))));
 
     wireMockRule.stubFor(
         get(urlPathEqualTo(path))
@@ -132,29 +164,88 @@ public class DiscoveryTest {
 
     List<String> versions = new ArrayList<>();
     versions.add(version);
-    Set<APIResource> apiResources = discovery.findAll(group, versions, version);
+    Set<APIResource> apiResources = discovery.findAll();
     wireMockRule.verify(1, getRequestedFor(urlPathEqualTo(path)));
     assertEquals(2, apiResources.size());
   }
 
   @Test
-  public void findAllShouldReturnEmptyListWhenApiResponseIsNotSuccess() {
+  public void findAllShouldThrowImcompleteDiscoveryExceptionWhenAtLeastOneResourceDiscoveryApiResponseIsNotSuccess() throws ApiException {
     Discovery discovery = new Discovery(apiClient);
 
-    String group = "test.com";
+    wireMockRule.stubFor(
+        get("/api")
+            .willReturn(
+                aResponse()
+                    .withStatus(200)
+                    .withHeader("Content-Type", "application/json")
+                    .withBody(
+                        apiClient
+                            .getJSON()
+                            .serialize(new V1APIVersions()))));
+
+    String groupSuccess = "test.com";
     String version = "v1";
-    String path="/apis/"+group+'/'+version;
+    String pathSuccess="/apis/"+groupSuccess+'/'+version;
+
+    String groupError = "testError.com";
+    String pathError="/apis/"+groupError+'/'+version;
 
     wireMockRule.stubFor(
-        get(urlPathEqualTo(path))
+        get("/apis")
+            .willReturn(
+                aResponse()
+                    .withStatus(200)
+                    .withHeader("Content-Type", "application/json")
+                    .withBody(
+                        apiClient
+                            .getJSON()
+                            .serialize(new V1APIGroupList()
+                                .addGroupsItem(new V1APIGroup()
+                                    .name(groupError)
+                                    .preferredVersion(new V1GroupVersionForDiscovery().version(version))
+                                    .versions(Arrays.asList(
+                                        new V1GroupVersionForDiscovery().version(version))))
+                                .addGroupsItem(new V1APIGroup()
+                                    .name(groupSuccess)
+                                    .preferredVersion(new V1GroupVersionForDiscovery().version(version))
+                                    .versions(Arrays.asList(
+                                        new V1GroupVersionForDiscovery().version(version))))))));
+
+    wireMockRule.stubFor(
+        get(urlPathEqualTo(pathError))
             .willReturn(
                 aResponse()
                     .withStatus(503)));
 
+    wireMockRule.stubFor(
+        get(urlPathEqualTo(pathSuccess))
+            .willReturn(
+                aResponse()
+                    .withStatus(200)
+                    .withHeader("Content-Type", "application/json")
+                    .withBody(
+                        apiClient
+                            .getJSON()
+                            .serialize(new V1APIResourceList()
+                                .resources(
+                                    Arrays.asList(
+                                      new V1APIResource()
+                                        .name("first"),
+                                      new V1APIResource()
+                                        .name("second")))))));
+
     List<String> versions = new ArrayList<>();
     versions.add(version);
-    Set<APIResource> apiResources = discovery.findAll(group, versions, version);
-    wireMockRule.verify(1, getRequestedFor(urlPathEqualTo(path)));
-    assertEquals(0, apiResources.size());
+    Set<APIResource> apiResources = null;
+    try{
+      discovery.findAll();
+      fail("Should have throw ImcompleteDiscoveryException");
+    } catch (IncompleteDiscoveryException e) {
+      apiResources = e.getDiscoveredResources();
+    }
+    wireMockRule.verify(1, getRequestedFor(urlPathEqualTo(pathError)));
+    wireMockRule.verify(1, getRequestedFor(urlPathEqualTo(pathSuccess)));
+    assertEquals(2, apiResources.size());
   }
 }

--- a/util/src/test/java/io/kubernetes/client/DiscoveryTest.java
+++ b/util/src/test/java/io/kubernetes/client/DiscoveryTest.java
@@ -28,6 +28,7 @@ import io.kubernetes.client.openapi.models.V1APIResourceList;
 import io.kubernetes.client.openapi.models.V1APIVersions;
 import io.kubernetes.client.util.ClientBuilder;
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Set;
@@ -129,7 +130,9 @@ public class DiscoveryTest {
                                       new V1APIResource()
                                         .name("second")))))));
 
-    Set<APIResource> apiResources = discovery.findAll(group, List.of(version), version);
+    List<String> versions = new ArrayList<>();
+    versions.add(version);
+    Set<APIResource> apiResources = discovery.findAll(group, versions, version);
     wireMockRule.verify(1, getRequestedFor(urlPathEqualTo(path)));
     assertEquals(2, apiResources.size());
   }
@@ -148,7 +151,9 @@ public class DiscoveryTest {
                 aResponse()
                     .withStatus(503)));
 
-    Set<APIResource> apiResources = discovery.findAll(group, List.of(version), version);
+    List<String> versions = new ArrayList<>();
+    versions.add(version);
+    Set<APIResource> apiResources = discovery.findAll(group, versions, version);
     wireMockRule.verify(1, getRequestedFor(urlPathEqualTo(path)));
     assertEquals(0, apiResources.size());
   }

